### PR TITLE
Split UnaryOps.CAST into CAST and BITCAST

### PR DESCRIPTION
**Pull request by wpmed92**
This PR introduces `UnaryOps.BITCAST`, and it is used instead of `UnaryOps.CAST` and bitcast=true boolean flag combination.

`a = Tensor([1,2,3], dtype=dtypes.float).bitcast(dtypes.int).realize()`

 before

```
  0 ━┳ STORE MemBuffer(idx=0, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))
  1  ┗━┳ CAST (dtypes.int, True)
  2    ┗━━ LOAD MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))
((LazyOp(op=BufferOps.STORE, src=(LazyOp(op=UnaryOps.CAST, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))),), arg=(dtypes.int, True)),), arg=MemBuffer(idx=0, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))),), [Opt(op=OptOps.LOCAL, axis=0, amt=3)])
```

after
```
 0 ━┳ STORE MemBuffer(idx=0, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))
 1  ┗━┳ BITCAST dtypes.int
 2    ┗━━ LOAD MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))
((LazyOp(op=BufferOps.STORE, src=(LazyOp(op=UnaryOps.BITCAST, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))),), arg=dtypes.int),), arg=MemBuffer(idx=0, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True),)))),), [Opt(op=OptOps.LOCAL, axis=0, amt=3)])
```

### CODE REVIEW

**Comment by geohot** [comment]
Good to see you back!

**Comment by geohot** [comment]
arg of cast should be changed to only be the dtype, we don't need the True/False anymore since CAST/BITCAST are different ops

**Comment by wpmed92** [comment]
`arg` for casts is now just the `dtype`. Two failing tests are due to old sops.gz, i’ll update it.

File: tinygrad/engine/schedule.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ def _recurse_lb(buf:LazyBuffer, realizes:Dict[LazyBuffer, None], allbufs:Dict[La
       if len(buf.st.views) == 1 and buf.st.views[-1].mask and all_int(buf.base.st.shape) and \
           prod(buf.base.st.shape) >= prod([y-x for x,y in buf.st.views[-1].mask]):
         simple_pads.add(buf.base)
-      elif buf.base.op is UnaryOps.CAST and isinstance(buf.base.srcs[0].dtype, ImageDType) and isinstance(buf.base.arg[0], ImageDType):
+      elif buf.base.op in [UnaryOps.CAST, UnaryOps.BITCAST] and isinstance(buf.base.srcs[0].dtype, ImageDType) and \
```
Comment (Qazalin) [inline tinygrad/engine/schedule.py:110]: I think bitcast doesn't apply to ImageDType. See https://github.com/tinygrad/tinygrad/blob/master/tinygrad/features/image.py

File: tinygrad/engine/schedule.py [hunk 2]
```diff
@@ -206,7 +207,7 @@ def _graph_schedule(outs:List[LazyBuffer], seen:Set[LazyBuffer]) -> Tuple[Defaul
           if not st.contiguous or tr_next.op in ReduceOps: break
           tr = tr_next
         # don't cast to higher size before store (tr cannot be realized if forced_realize)
-        if tr.op is UnaryOps.CAST and tr.arg[0].itemsize > tr.srcs[0].dtype.itemsize:
+        if tr.op in [UnaryOps.CAST, UnaryOps.BITCAST] and tr.arg.itemsize > tr.srcs[0].dtype.itemsize:
```
Comment (Qazalin) [inline tinygrad/engine/schedule.py:210]: bitcast doesn't apply, we don't support shape changing bitcast.

**Comment by wpmed92** [comment]
@Qazalin @geohot

- `arg` is now just the `dtype` (it's not a tuple anymore)
- no bitcast on `ImageDType`
- regenerated dataset so that `sops.gz` has the new `UnaryOps.CAST` and `UnaryOps.BITCAST` in the ASTs (for this I also had to update the UNROLL action in `search`, because I got a wrong action for `UNROLL, axis=4` for the new dataset)

File: tinygrad/codegen/kernel.py [hunk 3]
```diff
@@ -343,7 +343,7 @@ def _apply_tc_opt(self, use_tensor_cores:int, axis:int, opt_level:int) -> bool:
     if use_tensor_cores and self.opts.has_local and self.reduceop and self.reduceop.op is ReduceOps.SUM and self.opts.device in tensor_cores:
       for tc in tensor_cores[self.opts.device]:
         has_cast = tc.dtype_in != tc.dtype_out
-        if has_cast and not(self.reduceop.src[0].op is UnaryOps.CAST and self.reduceop.src[0].arg[0] == tc.dtype_out): continue
+        if has_cast and not(self.reduceop.src[0].op in [UnaryOps.CAST, UnaryOps.BITCAST] and self.reduceop.src[0].arg == tc.dtype_out): continue
```
Comment (Qazalin) [inline tinygrad/codegen/kernel.py:346]: tensor cores don't have bitcast.

File: tinygrad/lazy.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ def __init__(self, device:str, st:ShapeTracker, dtype:DType,
       self.op, self.arg, self.srcs = op, arg, srcs  # this is a LazyOp, except the src is LazyBuffers and not LazyOps
       assert self.op is not LoadOps.ASSIGN or srcs[1].base.realized is not None, "assign target must be realized"
 
-      if (self.op is LoadOps.CONTIGUOUS or (self.op is UnaryOps.CAST and self.arg[1] is True)) and srcs[0].st.consecutive and \
+      if (self.op is LoadOps.CONTIGUOUS or (self.op is UnaryOps.BITCAST)) and srcs[0].st.consecutive and \
```
Comment (Qazalin) [inline tinygrad/lazy.py:37]: nit: redundant `(`

File: tinygrad/codegen/kernel.py [hunk 5]
```diff
@@ -352,7 +352,8 @@ def buf_index(src: LazyOp) -> Optional[int]:
           # TODO: apply tc even if the sources are not from LOAD
           if src.op is BufferOps.LOAD and src.arg.dtype == tc.dtype_in: return self.bufs.index(cast(MemBuffer, src.arg))
           try:
-            if opt_level >= 1 and src.op is UnaryOps.CAST and src.arg[0] == tc.dtype_in: return self.bufs.index(cast(MemBuffer, src.src[0].arg))
+            if opt_level >= 1 and src.op in [UnaryOps.CAST, UnaryOps.BITCAST] and src.arg == tc.dtype_in:
```
Comment (Qazalin) [inline tinygrad/codegen/kernel.py:355]: this is tc too...

Comment (wpmed92) [reply tinygrad/codegen/kernel.py:355]: In the old version we didn’t filter based on cast arg true/false, so didn’t CAST also mean bitcast?

Comment (wpmed92) [reply tinygrad/codegen/kernel.py]: Just fixed it.

File: tinygrad/codegen/kernel.py [hunk 6]
```diff
@@ -352,7 +352,8 @@ def buf_index(src: LazyOp) -> Optional[int]:
           # TODO: apply tc even if the sources are not from LOAD
           if src.op is BufferOps.LOAD and src.arg.dtype == tc.dtype_in: return self.bufs.index(cast(MemBuffer, src.arg))
           try:
-            if opt_level >= 1 and src.op is UnaryOps.CAST and src.arg[0] == tc.dtype_in: return self.bufs.index(cast(MemBuffer, src.src[0].arg))
+            if opt_level >= 1 and src.op is UnaryOps.CAST and src.arg == tc.dtype_in:
+              return self.bufs.index(cast(MemBuffer, src.src[0].arg))
```
Comment (Qazalin) [inline tinygrad/codegen/kernel.py:356]: nit: why is this two lines?

Comment (wpmed92) [reply tinygrad/codegen/kernel.py:356]: Line was too long with the bitcast, will update it in a sec.

**Comment by Qazalin** [comment]
nice - new graph:
image

I'm not sure about the wrong UNROLL action though - What was the error?

**Comment by wpmed92** [comment]
I was getting this for `PYTHONPATH="." GPU=1 python3 extra/optimization/get_action_space.py`:

Screenshot 2024-05-13 at 09 32 48

**Comment by Qazalin** [comment]
did you try bumping the DB version for the actions issue? you might just need to revalidate the sqlite db instead of changing action space script.

**Comment by wpmed92** [comment]
Still wrong after db bump, only seems to be solved by bumping axis in unroll.

**Comment by chenyuxyz** [comment]
the error is saying the handcoded opt had used a action that's not in the search action lists. can you print the kernel that caused this?

**Comment by chenyuxyz** [comment]
congrats! george@tinygrad.org to claim

**Pull request merged**